keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,404 @@
+
+* As a developer I would like to rely on a self-healing platform for my stateful workloads
+
+This KEP plans to introduce the following CSI capabilities.
+
+* In the [CSI spec](https://github.com/container-storage-interface/spec/blob/master/spec.md), introduce `SAFE_DETACH` capablity in both controller and node service capabilities.
+
+* In the [CSIDriverSpec](https://github.com/kubernetes/kubernetes/blob/v1.20.2/pkg/apis/storage/types.go#L266), add a `SafeDetach` boolean to indicate if it is safe to detach.
+
+```go
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+  ...
+
+  // SafeDetach indicates this CSI volume driver is able to perform detach
+  // operations, when set to true, the CSI volume driver needs to ensure that 
+  // ControllerUnpublishVolume() implementation is checking for detach safety
+  // +optional
+  SafeDetach *bool
+}
+
+```
+
+When a node is shut down, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.
+
+After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.
+
+The Pod GC Controller, part of the kube-controller-manager, would need to go through all the Pods in terminating state, verify that the Node is NotReady, and then check if the `SafeDetach` flag in CSIDriver is set to true. If so, forcefully delete pods that are stuck.
+
+Specifically the Pod GC Controller (the only controller able to forcefully delete pods) would also check now for stateful pods, and select those that satisfy ALL of the following conditions:
+
+- pods that are backed by a provisioner that support safe detach
+- pods that are backed by volumes that aren't marked as `ReadWriteMany, this would avoid breaking having multiple replicas of the same pod.
+
+Once pods are selected and forcefully deleted, the attachdetach reconciler should check if the `SAFE_DETACH` CSI controller capability is true. If true, it should skip the `attachedVolume.MountedByNode` check for the volumes backing these pods and allow `ControllerUnpublishVolume` to happen before `NodeUnpublishVolume` and/or `NodeUnstageVolume` are called.
+
+This would trigger the deletion of the `volumeAttachement` objects, which would in turn trigger `ControllerUnpublishVolume()`. The provider would need to ensure there that no volume is being used through a mount point. If unable to determine the usage, the storage provider can return an error without detaching the volume, the `external-attacher` would then retry with backoff.
+
+The Pod GC Controller would also apply a `quarantine` taint on the nodes whose pods are forcefully deleted due the shutdown.
```

Comment:
@yastij wrote below on 18 Jul 2019 below:

> This would be part of a collection of 3 KEPs:
> 1 - node failover (this one)
> 2 - node GC
> 3 - statefulset evictions using NoExecute

why apply a `quarantie` taint instead of `NoExecute` taint?

Reply:
Can you provide a link to the 3rd KEP "statefulset evictions using NoExecute"?

Reply:
`NoExecute` is to make sure scheduler does not schedule additional pods to the node. This new taint `quarantine` is to differentiate it from the regular "do not schedule" case.  It indicates the node also needs to be "cleaned up" before new pods can be scheduled on it.

Reply:
there is no  3rd KEP "statefulset evictions using NoExecute" yet. This is mentioned https://github.com/kubernetes/enhancements/pull/1116#issuecomment-512840909

Reply:
We need to create KEP 2, 3 as soon as possible?